### ISSUE TITLE
Serious issue with subtrees!

### ISSUE THREAD

**Issue by fpug** [caesar0301/treelib#47]
Hi guys,

I was trying your library out, it looks great.
Yet I believe there is a design issue with how you implemented the subtrees.  Right now a tree is defined by a dictionary of nodes and the id of the root node, the hierarchy being expressed in the bpointer and fpointer of the individual node instances.

What happens when I create a subtree?  A new tree object is created, that has a subset of the nodes of the original tree.  As you mention the copy is shallow, so the nodes themselves are not deep copied. This is useful, because in principle it allows to operate on the subtrees influencing the structure of the main tree, and vice-versa.

But what if I first create a subtree out of a tree, then add a node either to this subtree or to the original tree?  What happens is that the fpointers and bpointers of the nodes in question will be updated to include the new node, but _just the nodes dictionary of either the tree or subtree will be updated_.

So when for example the function _print_backend(...) gets invoked on the tree if I modified the subtree, or on the subtree if I modified the tree, it will throw an error at the line:

``` python
queue = [self[i] for i in self[nid].fpointer if filter(self[i])]
```

because there will be identifiers that are in the list of fpointers but NOT in the nodes of the tree/subtree.  

How would you go about fixing that?

Have a great day and thanks,

Francesco

**Comment by caesar0301** [comment]
Yeah, you got it. Its a devil hidden in the implementation. We should find a good solution to work around this when we keep the following principle:
* The shallow copy (subtree) should keep consistence with its source tree.

This meas the original tree and subtree should keep a bidirectional communication to know the update of each other. One potential solution in my mind is to keep the all nodes seen by the original tree and its subtrees. i.e., they share the same view of all nodes.

Any better ideas?

Xiaming

**Comment by fpug** [comment]
What if we implement shallow subtrees as "views" of the original tree rather than as full-blown trees?
Deep copies would still get you an actual tree object.

**Comment by caesar0301** [comment]
That's right! Its even much better if they (tree and its subtree) share the same class type. That is, the subtree method should return a Tree object.

Could you implement your view-idea as a patch and we improve it? Thanks

**Comment by brainysmurf** [comment]
I have a subtree with 200,000 leaves and it took several minutes to complete, which led me to posting this message as well. Not sure it has to be a Tree object, though, maybe a class with compatible method names? I'd be happy to take a deeper look.

**Comment by c0fec0de** [comment]
I think anytree (http://anytree.readthedocs.io/en/latest/) should solve the subtree issue. It does not distinguish between Tree objects and Node objects. There are just nodes. A Node without a parent is the root of a tree. So creating a subtree is just detaching a node from its parent by setting the .parent attribute to None. For attach just set the parent attribute to another node. There are various attributes like .children etc. which are updated automatically inside. Please see the documentation.

If you want to iterate over a subtree or render it, you do not need to modify the tree. Iteration, Rendering and Walking are just strategies applied to the tree. You have to specify the starting node, which can be a root or any other node.

The internal treelib data structure conflicts with the approach above. 

HTH
c0fec0de

**Comment by leonardbinet** [comment]
This issue will be solved in v1.6.0 relase: https://github.com/caesar0301/treelib/issues/128 notably with this improvement: https://github.com/caesar0301/treelib/issues/127.
